core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1524,19 +1538,62 @@ public void waitForOpsToComplete(final long seqNo) throws InterruptedException {
      * {@link org.elasticsearch.index.seqno.GlobalCheckpointTracker#markAllocationIdAsInSync(String, long)}
-     * for additional details.
+     * for additional details. Because this operation could be completed asynchronously on another thread, the caller must provide a latch;
+     * this latch will be counted down after the operation completes.
      *
-     * @param allocationId    the allocation ID of the shard to mark as in-sync
-     * @param localCheckpoint the current local checkpoint on the shard
+     * @param allocationId      the allocation ID of the shard to mark as in-sync
+     * @param localCheckpoint   the current local checkpoint on the shard
+     * @param latch             a latch that is counted down after the operation completes
+     * @param onFailureConsumer a callback if an exception occurs completing the operation
      */
-    public void markAllocationIdAsInSync(final String allocationId, final long localCheckpoint) throws InterruptedException {
-        verifyPrimary();
-        getEngine().seqNoService().markAllocationIdAsInSync(allocationId, localCheckpoint);
+    public void markAllocationIdAsInSync(
+            final String allocationId,
+            final long localCheckpoint,
+            final CountDownLatch latch,
+            final Consumer<Exception> onFailureConsumer) {
         /*
-         * We could have blocked waiting for the replica to catch up that we fell idle and there will not be a background sync to the
-         * replica; mark our self as active to force a future background sync.
+         * Before marking the shard as in-sync we acquire an operation permit. We do this so that there is a barrier between marking a shard
+         * as in-sync and relocating a shard. If we acquire the permit then no relocation handoff can complete before we are done marking
+         * the shard as in-sync. If the relocation handoff holds all the permits then after the handoff completes and we acquire the permit
+         * then the state of the shard will be relocated and this recovery will fail.
          */
-        active.compareAndSet(false, true);
+        acquirePrimaryOperationPermit(
+                new ActionListener<Releasable>() {
+                    @Override
+                    public void onResponse(final Releasable releasable) {
+                        // we could have been relocated while waiting for a permit
+                        if (state() == IndexShardState.RELOCATED) {
+                            onFailure(new IndexShardRelocatedException(shardId));
+                        }
+                        boolean success = false;
+                        try {
+                            getEngine().seqNoService().markAllocationIdAsInSync(allocationId, localCheckpoint);
+                            /*
+                             * We could have blocked so long waiting for the replica to catch up that we fell idle and there will not be a
```

Comment:
can you explain we this is needed? so we don't have the background sync? what happens?